k2/csrc/weights.cc

```diff
@@ -68,11 +69,81 @@ WfsaWithFbWeights::WfsaWithFbWeights(const Fsa &fsa, const float *arc_weights,
     : fsa(fsa), arc_weights(arc_weights), weight_type(t) {
   if (IsEmpty(fsa)) return;
   CheckInput(fsa, arc_weights);
+  ComputeForwardWeights();
+  ComputeBackardWeights();
+}
+
+// Mohri, M. 2002. Semiring framework and algorithms for shortest-distance
+// problems, Journal of Automata, Languages and Combinatorics 7(3): 321-350,
+// 2002.
+void WfsaWithFbWeights::ComputeForwardWeights() {
+  auto num_states = fsa.NumStates();
+  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
```

Comment:
I prefer to use
```
std::vector<double> forward_state_weights;
```

and
```
forward_state_weights.resize(num_states);
```

Reply:
I'm sort of fine with this, but there is an interface issue, users will get a vector instead of a raw pointer, so for our other functions, we may need to make them accept vector (instead of pointer) of weights.

Reply:
`*_state_weights` are private data members, can we return `*_state_weights.data()`
instead of `*_state_weights.get()`?

Users can still get a pointer.

Reply:
but that will be: we use vector as a pointer only for the reason that we want its ability for memory managment..that is what `unique_ptr` doing..

Reply:
Another advantage of using `std::vector` here:

you can use `resize()` inside `ComputeForwardWeights`, which **may** reuse the memory
allocated last time.

`std::unique_ptr` here always results in memory allocation each time `ComputeForwardWeights` is called; I guess it will be called multiple times
and that's why Dan worries about its efficiency.

Reply:
`std::vector` has its own memory allocator, which has better support for cache.

Reply:
mm.. the input fsa is const, so we will definitly create `state_weight` only one time in the constructor.. there's no effciency issue before the loop.

Reply:
> Another advantage of using `std::vector` here:
> 
> you can use `resize()` inside `ComputeForwardWeights`, which **may** reuse the memory
> allocated last time.
> 
> `std::unique_ptr` here always results in memory allocation each time `ComputeForwardWeights` is called; I guess it will be called multiple times
> and that's why Dan worries about its efficiency.

ComputeForwardWeights() won't be called more than once for any given object.  It can only be called from the constructor.  From a memory allocation point of view I don't think std::vector and unique_ptr really differ, and I don't mind.